refresh.template.py

```diff
@@ -147,14 +148,30 @@ def _get_headers_msvc(compile_args: typing.List[str], source_path: str):
         # End:   template filled by Bazel
     ))
 
-    header_search_process = subprocess.run(
-        header_cmd,
-        stderr=subprocess.PIPE,
-        stdout=subprocess.DEVNULL,
-        env=environment,
-        encoding=locale.getpreferredencoding(),
-        check=False, # We explicitly ignore errors and carry on.
-    )
+    def _search_headers(command):
+        return subprocess.run(
+            command,
+            stderr=subprocess.PIPE,
+            stdout=subprocess.DEVNULL,
+            env=environment,
+            encoding=locale.getpreferredencoding(),
+            check=False, # We explicitly ignore errors and carry on.
+        )
+
+    header_search_process = _search_headers(header_cmd)
+    if header_search_process.returncode == 1:
+        # Write header_cmd to a temporary file, so we can use it as a parameter file to cl.exe, because Windows cmd has a limitation of 8191 charactors.
+        # See https://docs.microsoft.com/en-us/troubleshoot/windows-client/shell-experience/command-line-string-limitation
+        # To overcome this issue, we can call command parameters from a params file and use the '@' switch to pass the file name.
+        # E.g. cl.exe @params_file.txt
+        # If the return code is 1, it means the command line is too long (>=8191), so we write the command to a temp file and call from it.
+        temp_params = tempfile.NamedTemporaryFile('w')
+        # should skip cl.exe the 1st line
+        temp_params.write('\n'.join(header_cmd[1:]))
+        header_cmd = [header_cmd[0], f'@{temp_params}']
+        header_search_process = _search_headers(header_cmd)
+        # close and delete the temp file we created
+        temp_params.close()
```

Comment:
Thanks for documenting things so well, and sharing code via the local fn!

A few last things:
1) I didn't read the docs carefully enough on NamedTemporaryFile the first time. I'm now seeing 
>Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows NT or later).


...which seems like it would cause problems. How do you think we should solve this? (Perhaps we do need to use tempfile.mkstemp after all...sorry.)
2) I think we want to be careful to make sure that we're flushing the write offer before calling subprocess.

Reply:
The doc in the source code says that NamedTemporaryFile does use mkstemp to create file... I'd like to double check that

Reply:
```python
def NamedTemporaryFile(mode='w+b', buffering=-1, encoding=None,
                       newline=None, suffix=None, prefix=None,
                       dir=None, delete=True, *, errors=None):
    """Create and return a temporary file.
    Arguments:
    'prefix', 'suffix', 'dir' -- as for mkstemp.
    'mode' -- the mode argument to io.open (default "w+b").
    'buffering' -- the buffer size argument to io.open (default -1).
    'encoding' -- the encoding argument to io.open (default None)
    'newline' -- the newline argument to io.open (default None)
    'delete' -- whether the file is deleted on close (default True).
    'errors' -- the errors argument to io.open (default None)
    The file is created as mkstemp() would do it.

    Returns an object with a file-like interface; the name of the file
    is accessible as its 'name' attribute.  The file will be automatically
    deleted when it is closed unless the 'delete' argument is set to False.
    """

    prefix, suffix, dir, output_type = _sanitize_params(prefix, suffix, dir)

    flags = _bin_openflags

    # Setting O_TEMPORARY in the flags causes the OS to delete
    # the file when it is closed.  This is only supported by Windows.
    if _os.name == 'nt' and delete:
        flags |= _os.O_TEMPORARY

    (fd, name) = _mkstemp_inner(dir, prefix, suffix, flags, output_type)
    try:
        file = _io.open(fd, mode, buffering=buffering,
                        newline=newline, encoding=encoding, errors=errors)

        return _TemporaryFileWrapper(file, name, delete)
    except BaseException:
        _os.unlink(name)
        _os.close(fd)
        raise

def _mkstemp_inner(dir, pre, suf, flags, output_type):
    """Code common to mkstemp, TemporaryFile, and NamedTemporaryFile."""

    names = _get_candidate_names()
    if output_type is bytes:
        names = map(_os.fsencode, names)

    for seq in range(TMP_MAX):
        name = next(names)
        file = _os.path.join(dir, pre + name + suf)
        _sys.audit("tempfile.mkstemp", file)
        try:
            fd = _os.open(file, flags, 0o600)
        except FileExistsError:
            continue    # try again
        except PermissionError:
            # This exception is thrown when a directory with the chosen name
            # already exists on windows.
            if (_os.name == 'nt' and _os.path.isdir(dir) and
                _os.access(dir, _os.W_OK)):
                continue
            else:
                raise
        return (fd, _os.path.abspath(file))

    raise FileExistsError(_errno.EEXIST,
                          "No usable temporary file name found")
```

Pasted the code here. It calls `_mkstemp_inner` which is same as `mkstemp`, and it uses a `_RandomNameSequence` to generate temp file name. `TMP_MAX` is 10000 to guarantee there's no name conflict.


Now the test code:
```python
import os
import tempfile
import unittest
from concurrent.futures.thread import ThreadPoolExecutor


def _create_temp(bin):
    with tempfile.NamedTemporaryFile('w') as f:
        bin.append(f.name)
        f.write('abc')


class TestNamedTemporaryFile(unittest.TestCase):
    def test_mt_create(self):
        TEMP_BIN = []
        Q = 1000000

        pool = ThreadPoolExecutor(os.cpu_count())
        [pool.submit(_create_temp, TEMP_BIN) for _ in range(Q)]
        pool.shutdown(wait=True)
        self.assertEqual(len(TEMP_BIN), Q)
        self.assertEqual(len(set(TEMP_BIN)), Q)


if __name__ == '__main__':
    unittest.main()
```
Which is passed here.

Reply:
Huh, interesting!

 I wonder what part prompted them to add that reopen warning to the docs.

Reply:
Aha! Looks like the thing that causes the problem with NamedTemporaryFile vs mkstemp in the source above is is O_TEMPORARY.

Explained over at  https://stackoverflow.com/questions/15169101/how-to-create-a-temporary-file-that-can-be-read-by-a-subprocess

And flushing comes up there, too.

---

I also want to make sure we aren't accidentally waiting on each other. I'm still waiting on you fixing this one before merging, right? And you're not waiting on me?

Reply:
@cpsauer Ah, I got what you mean here, and there's indeed a potentail risk for `cl.exe` to read an opened temp file (while it works for me though...). Let's delete it manually as this post did.